lib/recursive-methods.rb

```diff
@@ -1,49 +1,107 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    t_fact(n, 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def t_fact(n, i)
+    if n < 0
+        raise ArgumentError, "number is less than zero"
+    else
+        return i if n == 1 || n == 0
+    end
+
+    return t_fact(n - 1, i * n)
+end
+
+# Time complexity: ?O(n)
+# Space complexity: ? O(n^2)
 def reverse(s)
```

Comment:
👍 However this is O(n^2) for both time and space complexity due to the slice here.